README.md
# k-vigicrues

[![Latest Release](https://img.shields.io/github/v/tag/kalisio/k-vigicrues?sort=semver&label=latest)](https://github.com/kalisio/k-vigcrues/releases)
[![Build Status](https://app.travis-ci.com/kalisio/k-vigicrues.svg?branch=master)](https://app.travis-ci.com/kalisio/k-vigicrues)

A [Krawler](https://kalisio.github.io/krawler/) based service to download data from French flood warning system [Vigicrues](https://www.vigicrues.gouv.fr/)

To get support from **Vigicrues**, use the following contact address: <vigicrues@developpement-durable.gouv.fr>

## Description

The **k-vigicrues** job allow to scrape flood alerts from the following services: [https://www.vigicrues.gouv.fr/services/1/](https://www.vigicrues.gouv.fr/services/1/). The downloaded data are stored in a [MongoDB](https://www.mongodb.com/) database and more precisely in the collection `vigicrues`.

All records are stored in [GeoJson](https://fr.wikipedia.org/wiki/GeoJSON) format.

The job is executed according a specific cron expression. By default every 3 hours.

## Configuration

| Variable | Description |
|--- | --- |
| `DB_URL` | The mongoDB database URL. The default value is `mongodb://127.0.0.1:27017/vigicrues`. |
| `DEBUG` | Enables debug output. Set it to `krawler*` to enable full output. By default it is undefined. |

## Deployment

We personally use [Kargo](https://kalisio.github.io/kargo/) to deploy the service.

## Contributing

Please refer to [contribution section](./CONTRIBUTING.md) for more details.

## Authors

This project is sponsored by 

![Kalisio](https://s3.eu-central-1.amazonaws.com/kalisioscope/kalisio/kalisio-logo-black-256x84.png)

## License

This project is licensed under the MIT License - see the [license file](./LICENSE) for details





CHANGELOG.md
Please refer to GitHub issues for changelog starting from v1.6: https://github.com/kalisio/k-vigicrues/issues.

### 1.5.0 (2023-03-09)

##### New Features

*  Upgrade NodeJS to v16 and krawler to latest version (close [#31](https://github.com/kalisio/k-vigicrues/pull/31)) ([5dcc0cbc](https://github.com/kalisio/k-vigicrues/commit/5dcc0cbc786d038b1f205ae21c79b603d774332e))

##### Other Changes

*  Upgrade NodeJS to v16 and krawler to latest version [#31](https://github.com/kalisio/k-vigicrues/pull/31) ([28206f0e](https://github.com/kalisio/k-vigicrues/commit/28206f0e049af06b7c2502a705cab565f8252cc2))
*  Job fails due to invalid geometries - improved reliability ([#30](https://github.com/kalisio/k-vigicrues/pull/30)) ([59c56b63](https://github.com/kalisio/k-vigicrues/commit/59c56b63fde4f7f753bdf5329d32dffe5147ff58))
* //github.com/kalisio/k-vigicrues/issues/29 (add missing properties for sections/forecasts to make kano tooltip work as expected) ([c4dfe98e](https://github.com/kalisio/k-vigicrues/commit/c4dfe98e5d1a4d8486507033b50322d131407f10))
* //github.com/kalisio/k-vigicrues/issues/29 (filter properties for sections) ([722c47a2](https://github.com/kalisio/k-vigicrues/commit/722c47a2fabb8382cb69cc9f6fc8d28f981b2415))
* //github.com/kalisio/k-vigicrues/issues/29 ([bcf2dff5](https://github.com/kalisio/k-vigicrues/commit/bcf2dff5cc5298c850242e57ebe3741f9def8866))
* //github.com/kalisio/k-vigicrues/issues/28 ([84a7f77e](https://github.com/kalisio/k-vigicrues/commit/84a7f77e5f40463a6600bba8820701b47737221b))
* //github.com/kalisio/k-vigicrues/issues/29 ([29b609a8](https://github.com/kalisio/k-vigicrues/commit/29b609a886c2d6856f968afa96b41435c1af8ecd))
*  updated badge url ([4bfcd4a4](https://github.com/kalisio/k-vigicrues/commit/4bfcd4a421ac082f064a1bb0724cc7f9de5a271b))

### 1.4.0 (2020-12-22)

##### New Features

 * Handle the new version of the API (close [#29](https://github.com/kalisio/k-vigicrues/pull/29))
 * Allow to configure the returned geometry (close [#28](https://github.com/kalisio/k-vigicrues/pull/28))

##### Other Changes

* //github.com/kalisio/k-vigicrues/issues/29 (add missing properties for sections/forecasts to make kano tooltip work as expected) ([c4dfe98e](https://github.com/kalisio/k-vigicrues/commit/c4dfe98e5d1a4d8486507033b50322d131407f10))
* //github.com/kalisio/k-vigicrues/issues/29 (filter properties for sections) ([722c47a2](https://github.com/kalisio/k-vigicrues/commit/722c47a2fabb8382cb69cc9f6fc8d28f981b2415))
* //github.com/kalisio/k-vigicrues/issues/29 ([bcf2dff5](https://github.com/kalisio/k-vigicrues/commit/bcf2dff5cc5298c850242e57ebe3741f9def8866))
* //github.com/kalisio/k-vigicrues/issues/28 ([84a7f77e](https://github.com/kalisio/k-vigicrues/commit/84a7f77e5f40463a6600bba8820701b47737221b))
* //github.com/kalisio/k-vigicrues/issues/29 ([29b609a8](https://github.com/kalisio/k-vigicrues/commit/29b609a886c2d6856f968afa96b41435c1af8ecd))
*  updated badge url ([4bfcd4a4](https://github.com/kalisio/k-vigicrues/commit/4bfcd4a421ac082f064a1bb0724cc7f9de5a271b))

### 1.3.0 (2020-10-01)

##### Other Changes

* Updated release scripts

#### 1.2.1 (2020-09-14)

##### Bug Fixes

* NivSituVigiCruEnt property appears in sections

### 1.2.0 (2020-09-08)

##### New Features

*  Execute the job automatically when the the container is launched (close [#26](https://github.com/kalisio/k-vigicrues/pull/26)) ([5270cedd](https://github.com/kalisio/k-vigicrues/commit/5270ceddabaef3dce744bff77c717894c3211a20))
*  Archive the data (close [#23](https://github.com/kalisio/k-vigicrues/pull/23)) ([12c8ae7b](https://github.com/kalisio/k-vigicrues/commit/12c8ae7b71673533d5f744c5066b965a48ff90d3))
*  Allow to configure the cron expression [#25](https://github.com/kalisio/k-vigicrues/pull/25) ([fc679fd3](https://github.com/kalisio/k-vigicrues/commit/fc679fd32eaf4f5798381ae55cd82352287a3d2f))

### 1.1.0 (2020-04-29)

##### New Features

*  Upgrade to Node 12 (close [#24](https://github.com/kalisio/k-vigicrues/pull/24)) ([85fdc361](https://github.com/kalisio/k-vigicrues/commit/85fdc36152e0ab9cef5406e3178b90491c1865dd))

## 1.0.0 (2019-12-12)

##### New Features

*  Add Changelog (closed [#21](https://github.com/kalisio/k-vigicrues/pull/21)) ([1e5dee04](https://github.com/kalisio/k-vigicrues/commit/1e5dee042f0cc653c357232b668d9bc7c81adce8))
*  Add Changelog (closed 21) ([7b15926c](https://github.com/kalisio/k-vigicrues/commit/7b15926c2ee0840c9da416b7b33c8e5584e23af1))
*  Remove stations and observations jobs (closed [#20](https://github.com/kalisio/k-vigicrues/pull/20)) ([7eb524fb](https://github.com/kalisio/k-vigicrues/commit/7eb524fb9f441a15f86c46305595094d770a6cf9))
*  Remove unused fields from the observations (closed [#14](https://github.com/kalisio/k-vigicrues/pull/14)) ([b9923758](https://github.com/kalisio/k-vigicrues/commit/b9923758d664a3ffb1557c255217cf2ddf2df572))
*  Allow to configure the expiration period (closed [#16](https://github.com/kalisio/k-vigicrues/pull/16)) ([ea2f70ed](https://github.com/kalisio/k-vigicrues/commit/ea2f70ed5050b1f5a7a3a198e2eaa72493040885))
*  Allow to configure the colors of the sections (closed [#15](https://github.com/kalisio/k-vigicrues/pull/15)) ([81caf1d3](https://github.com/kalisio/k-vigicrues/commit/81caf1d38ac54ff8150d5a62e69f8d848be81feb))
*  Provide a property name to be used for display (closed [#13](https://github.com/kalisio/k-vigicrues/pull/13)) ([14405726](https://github.com/kalisio/k-vigicrues/commit/1440572658fc7d7eb7b5d3c643e87c5986e483a8))

##### Bug Fixes

*  Name property in the observations collection is null (closed [#18](https://github.com/kalisio/k-vigicrues/pull/18)) ([c379714c](https://github.com/kalisio/k-vigicrues/commit/c379714c7a1cfe408e23019cbb18f4cb09f5bf64))
*  Observations should handle the name property (closed [#17](https://github.com/kalisio/k-vigicrues/pull/17)) ([eb056219](https://github.com/kalisio/k-vigicrues/commit/eb056219e5e94853eb1cb97254ea87011fe6050d))

jobfile.js
import _  from 'lodash'
import moment from 'moment'
import { featureEach, getType, multiLineString, getCoords, cleanCoords, envelope, flatten } from '@turf/turf'
import makeDebug  from 'debug'

const debug = makeDebug('k-vigicrues')

const dbUrl = process.env.DB_URL || 'mongodb://127.0.0.1:27017/vigicrues'
const ttl = +process.env.TTL || (7 * 24 * 60 * 60)  // duration in seconds

export default {
  id: 'vigicrues',
  store: 'memory',
  options: {
    workersLimit: 1,
    faultTolerant: true
  },
  tasks: [{
    id: 'vigicrues',
    type: 'http',
    options: {
      url: 'https://www.vigicrues.gouv.fr/services/1/InfoVigiCru.geojson'
    }
  }],
  hooks: {
    tasks: {
      after: {
        readJson: {
        },
        generateStations: {
          hook: 'apply',
          function: (item) => {
            const features = _.get(item, 'data.features', [])
            let validFeatures = []
            _.forEach(features, feature => {
              // Ensure clean geometry as some line strings have degenerated lines
              if (getType(feature) === 'MultiLineString') {
                let validLines = []
                let nbInvalidGeometries = 0
                featureEach(flatten(feature), line => {
                  try {
                    cleanCoords(line, { mutate: true })
                    validLines.push(getCoords(line))
                  } catch (error) {
                    nbInvalidGeometries++
                  }
                })
                if (nbInvalidGeometries > 0) debug(`Filtering ${nbInvalidGeometries} invalid line(s) for ${feature.properties.LbEntCru}`)
                // Rebuild geometry from the clean line
                feature.geometry = multiLineString(validLines).geometry
                validFeatures.push(feature)
              } else if (getType(feature) === 'LineString')  {
                try {
                  cleanCoords(feature, { mutate: true })
                  validFeatures.push(feature)
                } catch (error) {
                  debug(`Filtering invalid line for ${feature.properties.LbEntCru}`)
                }
              }
              // Convert ID to numeric value
              _.set(feature, 'properties.gid', _.toNumber(feature.properties.gid))
              // Remove unused ID
              _.unset(feature, 'id')
              _.set(feature, 'properties.name', feature.properties.LbEntCru) // needed for timeseries
              _.set(feature, 'properties.NomEntVigiCru', feature.properties.LbEntCru) // backward compatibility
            })
            _.set(item, 'data.features', validFeatures)
          }
        },
        writeSections: {
          hook: 'updateMongoCollection',
          collection: 'vigicrues-sections',
          filter: { 'properties.gid': '<%= properties.gid %>' },
          upsert : true,
          transform: {
            pick: [
              'type',
              'geometry',
              'properties.gid',
              'properties.name',
              'properties.NomEntVigiCru',
              'properties.CdEntCru',
              'properties.CdTCC'
            ],
            inPlace: false
          },
          chunkSize: 256
        },
        generateForecasts: {
          hook: 'apply',
          function: (item) => {
            let forecastFeatures = []
            const features = _.get(item, 'data.features', [])
            _.forEach(features, feature => {
              let forecastFeature = envelope(feature)
              _.set(forecastFeature, 'time', moment.utc().toDate())
              _.set(forecastFeature, 'properties.gid', feature.properties.gid) // needed for timeseries
              _.set(forecastFeature, 'properties.name', feature.properties.LbEntCru) // needed for timeseries
              _.set(forecastFeature, 'properties.NivSituVigiCruEnt', feature.properties.NivInfViCr) // needed for timeseries
              _.set(forecastFeature, 'properties.NomEntVigiCru', feature.properties.LbEntCru) // backward compatibility
              forecastFeatures.push(forecastFeature)
            })
            item.data = forecastFeatures
          }
        },
        writeForecasts: {
          hook: 'writeMongoCollection',
          collection: 'vigicrues-forecasts',
        },
        clearData: {}
      }
    },
    jobs: {
      before: {
        createStores: { id: 'memory' },
        connectMongo: {
          url: dbUrl,
          // Required so that client is forwarded from job to tasks
          clientPath: 'taskTemplate.client'
        },
        createSectionsCollection: {
          hook: 'createMongoCollection',
          clientPath: 'taskTemplate.client',
          collection: 'vigicrues-sections',
          indices: [
            [{ 'properties.gid': 1 }, { unique: true }],
            { geometry: '2dsphere' }                                                                                                              
          ]
        },
        createForecastsCollection: {
          hook: 'createMongoCollection',
          clientPath: 'taskTemplate.client',
          collection: 'vigicrues-forecasts',
          indices: [
            [{ time: 1, 'properties.gid': 1 }, { unique: true }],
            { 'properties.NivSituVigiCruEnt': 1 },
            { 'properties.gid': 1, 'properties.NivSituVigiCruEnt': 1, time: -1 },
            [{ time: 1 }, { expireAfterSeconds: ttl }] // days in secs                                                                                                         
          ]
        }
      },
      after: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: ['memory']
      },
      error: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: ['memory']
      }
    }
  }
}
